CI: Add workflow for author approved label

ericphanson opened the pull request:
This intends to facilitate AutoMerge reading this label and bypassing some checks if the author approves. The intended use is:
1. AutoMerge would have guidelines that can be bypassed if the author approves. Without approval, these fail and the automerge comment instructs the author (well: whoever registered the version/package) to comment with "approved".
2. If the author does so, that triggers this workflow, which adds the "author-approved" label to the PR
3. The AutoMerge workflow would be updated to retrigger when this label is applied. It would run through the same logic, but this time notice the label, and use it to bypass those checks.

Currently this workflow is useless as AutoMerge does not have such conditional checks, but I wanted to put a PR to try it out.

Note that https://github.com/JuliaRegistries/RegistryCI.jl/pull/536 shows how easy it is to use label information within automerge itself, which I think makes this a viable route.

For the sake of testing, I will add text to trigger the regex:

- Created by: @ericphanson

ericphanson committed:
add workflow for author approved label

ericphanson commented:
approved

ericphanson committed:
add token/repo

ericphanson commented:
Ah, I can't test this here, since it's an [`issue_comment` triggered workflow](https://docs.github.com/en/actions/using-workflows/events-that-trigger-workflows#issue_comment):

> Note: This event will only trigger a workflow run if the workflow file is on the default branch.

I suppose if we merge this, I can test it in a separate PR that could presumably update the workflow and it would get the changes from the branch instead of from main?

DilumAluthge commented:
> I suppose if we merge this, I can test it in a separate PR that could presumably update the workflow and it would get the changes from the branch instead of from main?

I think, for issue comments and PR comments, it always picks up the workflow file from the default branch :(

We could make a fork of General and do some testing there?

DilumAluthge commented on .github/workflows/author_approval.yml:
```diff
@@ -0,0 +1,35 @@
+      - name: Verify package author
+        id: verify-author
+        env:
+          PR_BODY: ${{ github.event.issue.body }}
```
It might be good to add a comment here noting that this method (using intermediate environment variables) is a security recommendation from GitHub: https://docs.github.com/en/actions/security-guides/security-hardening-for-github-actions#using-an-intermediate-environment-variable

ericphanson replied:
```suggestion
          # We use an env variable, not direct interpolation, for security:
          # https://docs.github.com/en/actions/security-guides/security-hardening-for-github-actions#using-an-intermediate-environment-variable
          PR_BODY: ${{ github.event.issue.body }}
```

DilumAluthge commented on .github/workflows/author_approval.yml:
```diff
@@ -0,0 +1,35 @@
+          COMMENTER: ${{ github.event.comment.user.login }}
+        shell: julia --color=yes {0}
+        run: |
+          m = match(r"Created by: @([^\s]+)", ENV["PR_BODY"])
```
Do we need to enforce that at the end of the `@username`, we have either `\n` or `$`?

Otherwise, wouldn't an approval by user `@foo` be able to approve a PR created by user `@foobar`?

ericphanson replied:
I think it's fine actually:
```julia
julia> r = r"Created by: @([^\s]+)"
r"Created by: @([^\s]+)"

julia> match(r, "Created by: @foobar")[1]
"foobar"

julia> match(r, "Created by: @foo")[1]
"foo"
```
i.e. we will match the whole username, whatever it is, and then we string-compare (not regex) against the commenter

DilumAluthge commented:
I wonder if we could come up with a different name other than "author"? In the AutoMerge codebase, we always use "author" to refer to the actual GitHub PR author, so `@JuliaRegistrator`, `@jlbuild`, etc.

So it might be a little confusing to also use the same term "author" for the human that triggered Registrator.

Maybe "registration-trigger-user` or something like that?

ericphanson commented:
For the label? I think we should make sure it's understandable by the user, so they can know that their approval was picked up correctly. So I think it shouldn't be too obscure. What about `package-author-approved`?

ericphanson committed:
Update .github/workflows/author_approval.yml

ericphanson committed:
fixes from testing in other repo

ericphanson committed:
rename

ericphanson commented:
Ok, I've tested it [here](https://github.com/ericphanson/PkgA/pull/3), and updated this PR.

Mismatch:
<img width="664" alt="Screenshot 2024-01-14 at 23 48 40" src="https://github.com/JuliaRegistries/General/assets/5846501/2ef891f8-cfbf-499e-bcc9-273b9df831b3">
Match:
<img width="664" alt="Screenshot 2024-01-14 at 23 48 40" src="https://github.com/JuliaRegistries/General/assets/5846501/a3c2ca61-e866-4e6b-93c7-1176a57283af">

Note the times on the Julia part vary from 1-2s up to 15-20s, I'm not really sure why. However that code only runs when a comment is posted from a non-bot user that includes "approved", so hopefully it doesn't need to run too much.

DilumAluthge commented on .github/workflows/author_approval.yml:
```diff
@@ -0,0 +1,42 @@
+jobs:
+  label:
+    permissions:
+        pull-requests: write
```
Is there any narrower permission that we can give, other than `pull-requests: write`?

ericphanson replied:
I don't think there are narrower permissions: https://docs.github.com/en/actions/security-guides/automatic-token-authentication#permissions-for-the-github_token

ericphanson replied:
note however the token itself is only available to the final step of the workflow, which does not have any user-supplied inputs, so should be pretty secure

DilumAluthge commented on .github/workflows/author_approval.yml:
```diff
@@ -0,0 +1,42 @@
+    # 3) include the string "approved"
+    # If so, we will do the work to check that the commenter is the package author,
+    # and conditionally apply the author-approved label.
+    if: ${{ github.event.issue.pull_request && github.event.issue.user.type != 'Bot' && contains(github.event.comment.body, 'approved') }}
```
The word `approved` will need to appear with exact case matching (case sensitive) in the document, right? Should we document this somewhere?

ericphanson replied:
Yep, I think AutoMerge should mention it any time there’s a failure due to lack of approval (none exist yet). Perhaps once we have some we can add a faq entry as well. I don’t think it makes sense to add public docs with this PR as the mechanism isn’t used for anything yet.

DilumAluthge replied:
For now can you stick a comment in the YAML (right above this line, maybe) just pointing out that it's case sensitive?

ericphanson replied:
```suggestion
    # note: `approved` here is case-sensitive 
    if: ${{ github.event.issue.pull_request && github.event.issue.user.type != 'Bot' && contains(github.event.comment.body, 'approved') }}
```

DilumAluthge commented:
Can we also exclude the `github-actions[bot]` user?

ericphanson commented:
I have excluded all bot users

ericphanson committed:
Update .github/workflows/author_approval.yml

DilumAluthge commented:
> For the label? I think we should make sure it's understandable by the user, so they can know that their approval was picked up correctly. So I think it shouldn't be too obscure. What about `package-author-approved`?

Yeah let's go with that.

ericphanson committed:
Update author_approval.yml

ericphanson commented:
I had already renamed the label to `package-author-approved`. Just now, I changed the approval string to "merge approved", and realized in fact it is not case-sensitive, as `contains` is not. I updated the comment with a reference.

I think this is good to go, and am writing a PR to RegistryCI to make use of this in the sequential version skipping check (probably to the delight of like 2 people who like skipping versions). We can also use it for other things down the line.

DilumAluthge commented on .github/workflows/author_approval.yml:
```diff
@@ -0,0 +1,43 @@
+          COMMENTER: ${{ github.event.comment.user.login }}
+        shell: julia --compile=min --optimize=0 --color=yes {0}
+        run: |
+          m = match(r"Created by: @([^\s]+)", ENV["PR_BODY"])
```
Per the GitHub https://docs.github.com/en/enterprise-cloud@latest/admin/identity-and-access-management/iam-configuration-reference/username-considerations-for-external-authentication:

> Usernames for user accounts on GitHub.com can only contain alphanumeric characters and dashes (-).

So our regex only needs to check for alphanumeric characters (uppercase letters, lowercase letters, and numbers) and dashes.

Also, we can enforce that the "Created by" line either ends in a newline, or the end of the string.

```suggestion
          m = match(r"Created by: @([A-Za-z0-9]*+)(?:$|\n)", ENV["PR_BODY"])
```

DilumAluthge commented on .github/workflows/author_approval.yml:
```diff
@@ -0,0 +1,43 @@
+        if: ${{ steps.verify-author.outputs.verified == 'true' }}
+        env:
+          PR_NUM: ${{ github.event.issue.number }}
+          GH_TOKEN: ${{ secrets.GITHUB_TOKEN }}
```
```suggestion
          # We cannot use `${{ secrets.GITHUB_TOKEN }}` here, because
          # if we use `GITHUB_TOKEN` here, then the "label created" event
          # will not trigger any further GitHub Actions.
          GH_TOKEN: ${{ secrets.TAGBOT_TOKEN }}
```

DilumAluthge commented on .github/workflows/author_approval.yml:
```diff
@@ -0,0 +1,43 @@
+    # Run on comments that are:
+    # 1) on PRs, not issues,
+    # 2) not from bot users and,
+    # 3) include the string "merge approved"
```
```suggestion
    # 3) include the string "[merge approved]"
    # 4) include the string "[noblock]"
```

DilumAluthge commented on .github/workflows/author_approval.yml:
```diff
@@ -0,0 +1,43 @@
+    # If so, we will do the work to check that the commenter is the package author,
+    # and conditionally apply the author-approved label.
+    # note: `merge approved` here is NOT case-sensitive, see https://docs.github.com/en/actions/learn-github-actions/expressions#contains
+    if: ${{ github.event.issue.pull_request && github.event.issue.user.type != 'Bot' && contains(github.event.comment.body, 'merge approved') }}
```
```suggestion
    if: github.event.issue.pull_request && (github.event.issue.user.type != 'Bot') && (github.event.comment.body == '[merge approved]') && (github.event.comment.body == '[noblock]')
```

ericphanson replied:
This looks much more complicated, and I’m sure users will make mistakes. I’d much rather what I wrote and to change the RegistryCI check to not block on this phrase (which is already done in my PR)

DilumAluthge replied:
I don't love the special-casing in https://github.com/JuliaRegistries/RegistryCI.jl/pull/539.

It also seems simpler to me to tell users "always include `[noblock] in your comments", instead of saying "sometimes you need `[noblock]` in your comments, and other times you don't".

Can we just add text to the AutoMerge comment of the following form?

> If you are the author of this package, and you want to approve this PR and bypass the sequential-version-number check, please post a comment on this PR with the following text verbatim: `[merge approved] [noblock]`

ericphanson replied:
I think a longer comment means more chance to miss that text along with all the rest of it. It's already so long!

As a compromise, what about `[merge approved]` (with the square brackets), but without requiring `[noblock]`? Then the rule is "commands to AutoMerge occur in square brackets and comments with commands do not block automerge".

DilumAluthge replied:
I think I can live with that.

ericphanson replied:
```suggestion
    if: ${{ github.event.issue.pull_request && github.event.issue.user.type != 'Bot' && contains(github.event.comment.body, '[merge approved]') }}
```

DilumAluthge commented on .github/workflows/author_approval.yml:
```diff
@@ -0,0 +1,43 @@
+jobs:
+  label:
+    permissions:
+        pull-requests: write
```
```suggestion
        # We need `write` permissions on `pull-requests` in order to be able to
        # add/remove labels from PRs. As far as we can tell, there is no narrower
        # permission that we can use.
        pull-requests: write
```

DilumAluthge commented on .github/workflows/author_approval.yml:
```diff
@@ -0,0 +1,43 @@
+    permissions:
+        pull-requests: write
+    runs-on: ubuntu-latest
+    # Run on comments that are:
```
```suggestion
    # Run on comments that are satisfy all of the following:
```

DilumAluthge commented on .github/workflows/author_approval.yml:
```diff
@@ -0,0 +1,43 @@
+    runs-on: ubuntu-latest
+    # Run on comments that are:
+    # 1) on PRs, not issues,
+    # 2) not from bot users and,
```
```suggestion
    # 2) not from bot users
```

DilumAluthge commented on .github/workflows/author_approval.yml:
```diff
@@ -0,0 +1,43 @@
+    # 3) include the string "merge approved"
+    # If so, we will do the work to check that the commenter is the package author,
+    # and conditionally apply the author-approved label.
+    # note: `merge approved` here is NOT case-sensitive, see https://docs.github.com/en/actions/learn-github-actions/expressions#contains
```
```suggestion
    # note: `[merge approved]` here is NOT case-sensitive, see https://docs.github.com/en/actions/learn-github-actions/expressions#contains
```

DilumAluthge commented on .github/workflows/author_approval.yml:
```diff
@@ -0,0 +1,43 @@
+          PR_NUM: ${{ github.event.issue.number }}
+          GH_TOKEN: ${{ secrets.GITHUB_TOKEN }}
+          GH_REPO: ${{ github.repository }}
+        run: gh pr edit "$PR_NUM" --add-label "package-author-approved"
```
Can we rename the label from `package-author-approved` to `Override AutoMerge:package-author-approved`? That way, we'll have the convention that if a label is able to affect AutoMerge, then it starts with `Override AutoMerge:`?

ericphanson replied:
What about `Override AutoMerge: package author approved`? That matches better the other one, `Override AutoMerge: name similarity is okay`

DilumAluthge replied:
SGTM.

ericphanson replied:
```suggestion
        run: gh pr edit "$PR_NUM" --add-label "Override AutoMerge: package author approved"
```

ericphanson committed:
Apply suggestions from code review

Co-authored-by: Dilum Aluthge <dilum@aluthge.com>

ericphanson committed:
Apply suggestions from code review

Co-authored-by: Dilum Aluthge <dilum@aluthge.com>

ericphanson commented on .github/workflows/author_approval.yml:
```diff
@@ -0,0 +1,50 @@
+    # 1) on PRs, not issues,
+    # 2) not from bot users
+    # 3) include the string "[merge approved]"
+    # 4) include the string "[noblock]"
```
```suggestion
```

ericphanson committed:
Update .github/workflows/author_approval.yml

DilumAluthge commented on .github/workflows/author_approval.yml:
```diff
@@ -0,0 +1,49 @@
+          # will not trigger any further GitHub Actions.
+          GH_TOKEN: ${{ secrets.TAGBOT_TOKEN }}
+          GH_REPO: ${{ github.repository }}
+        run: gh pr edit "$PR_NUM" --add-label "Override AutoMerge: package author approved"
```
```suggestion
        run: gh pr edit "${PR_NUM:?}" --add-label "Override AutoMerge: package author approved"
```

ericphanson committed:
Update .github/workflows/author_approval.yml

Co-authored-by: Dilum Aluthge <dilum@aluthge.com>
